vo_gpu_next: add --target-colorspace-hint-mode option

kasper93 opened the pull request.

kasper93 commented on video/out/vo_gpu_next.c:
```diff
@@ -1023,8 +1025,25 @@ static bool draw_frame(struct vo *vo, struct vo_frame *frame)
+                    .primaries = PL_COLOR_PRIM_DISPLAY_P3,
+                    .hdr = {
+                        .max_luma = 1000,
+                        .min_luma = 50,
```
It obviously meant to be 50 mcd/m2 I don't want to set infinite contrast, typical screen is not oled and even oleds report 0.2 nits

dyphire replied:
Also feels that 50 nit is too high.
I guess this fallback mechanism is to deal with samples like https://github.com/mpv-player/mpv/issues/16488, but I also have an HDR sample that corrupts metadata and it is really made at 10000 nit (crazy). This fallback will also ruin it, and I don't think there is a good way here.

dyphire replied:
> The whole point of this change is to not set stupid swapchain parameters that are not supported by any display anyway.

This convinced me.

kasper93 commented:
Next step would be to deprecate `--target-colorspace-hint=auto` and set it to `yes` by default. SDR is now handled mostly correct.

dyphire commented:
> Next step would be to deprecate `--target-colorspace-hint=auto` and set it to `yes` by default. SDR is now handled mostly correct.

No, the defaults of `--target-colorspace-hint` should not be changed. Especially under vulkan even if the monitor is in SDR mode as soon as the HDR source is played it tries to use the fallback mechanism, which produces completely corrupted output.

`--vo=gpu-next --gpu-api=vulkan --target-colorspace-hint=yes`:
![](https://github.com/user-attachments/assets/52e63805-0962-4ae3-a2d6-1e88be6e09ef)
It should never be set to a default value of `yes` unless the target colorspace metadata is correctly fetched for all graphics backend implementations for each platform.

dyphire commented:
> Report this to your GPU driver vendor. Sending SDR in PQ as in this screenshot should work just fine. Additionally, your driver shouldn't expose HDR surfaces if it cannot handle them correctly. I know it's DWM just clipping the output, you can try enabling ACM in display options.

I use the `srgb` color gamut when the display is in SDR mode, so ACM should not be turned on, that would make all the displays incorrect.

> On AMD Vulkan driver properly signal supported surfaces to avoid any bullshit.

Unfortunately the Nvidia driver screwed everything up again, even with the latest vulkan driver it has the same logging behavior in both HDR mode and SDR mode:
```
[   1.000][v][vo/gpu-next/libplacebo] Available surface configurations:
[   1.000][v][vo/gpu-next/libplacebo]     0: VK_FORMAT_B8G8R8A8_UNORM                 VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
[   1.000][v][vo/gpu-next/libplacebo]     1: VK_FORMAT_B8G8R8A8_SRGB                  VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
[   1.000][v][vo/gpu-next/libplacebo]     2: VK_FORMAT_R8G8B8A8_UNORM                 VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
[   1.000][v][vo/gpu-next/libplacebo]     3: VK_FORMAT_R8G8B8A8_SRGB                  VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
[   1.000][v][vo/gpu-next/libplacebo]     4: VK_FORMAT_R16G16B16A16_SFLOAT            VK_COLOR_SPACE_EXTENDED_SRGB_LINEAR_EXT
[   1.000][v][vo/gpu-next/libplacebo]     5: VK_FORMAT_A2B10G10R10_UNORM_PACK32       VK_COLOR_SPACE_HDR10_ST2084_EXT
[   1.000][v][vo/gpu-next/libplacebo]     6: VK_FORMAT_A2B10G10R10_UNORM_PACK32       VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
[   1.000][v][vo/gpu-next/libplacebo] Picked surface configuration 6: VK_FORMAT_A2B10G10R10_UNORM_PACK32 + VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
[   1.002][d][vo/gpu-next/libplacebo] Requested alpha compositing mode: VK_COMPOSITE_ALPHA_OPAQUE_BIT_KHR
[   1.002][d][vo/gpu-next/libplacebo] Requested surface transform: VK_SURFACE_TRANSFORM_IDENTITY_BIT_KHR
[   6.756][v][vo/gpu-next/libplacebo] Picked surface configuration 5: VK_FORMAT_A2B10G10R10_UNORM_PACK32 + VK_COLOR_SPACE_HDR10_ST2084_EXT
[   6.756][d][vo/gpu-next/libplacebo] Requested alpha compositing mode: VK_COMPOSITE_ALPHA_OPAQUE_BIT_KHR
[   6.756][d][vo/gpu-next/libplacebo] Requested surface transform: VK_SURFACE_TRANSFORM_IDENTITY_BIT_KHR
```

> Generally, yes, I will not change it like that, but it's not mpv fault this output is broken. And since Vulkan doesn't have native api to query display capabilities this is the proper way to handle this.
>

I don't disagree with the changes made by this pr, I just don't agree with changing the defaults for `target-colorspace-hint` option. Real world driver vendor screw things up all the time, let's keep the default parameter behavior for maximum compatibility.
Users can manually change the option parameters to choose what they want.

kasper93 commented on video/out/vo_gpu_next.c:
```diff
@@ -1023,8 +1027,27 @@ static bool draw_frame(struct vo *vo, struct vo_frame *frame)
+        const struct pl_color_space *source = &frame->current->params.color;
+        const struct pl_color_space *target = &target_csp;
+        hint = *(p->next_opts->target_hint_mode == 1 ? source : target);
+        if (pl_color_transfer_is_hdr(hint.transfer) && !pl_primaries_valid(&hint.hdr.prim)) {
+            // We are targeting HDR, merge source only if it is also HDR.
+            if (pl_hdr_metadata_contains(&source->hdr, PL_HDR_METADATA_HDR10))
+                pl_color_space_merge(&hint, source);
+            // If we are targeting HDR and neither source nor target has valid primaries,
+            // use some default values. To set the display in the reasonable HDR mode.
+            if (pl_color_transfer_is_hdr(hint.transfer) && !pl_primaries_valid(&hint.hdr.prim)) {
```
`pl_color_space_merge` sets unknown values in `hint` to ones from `source`. So it's not an else branch, because source may not have primaries, in which case we fallback to display-p3 below.

Honestly I'm still thinking about the details of it. Main question is about this display-p3 fall back, but I will reply in sfan's comment with details.

kasper93 commented on player/lua/stats.lua:
```diff
@@ -888,7 +888,21 @@ local function append_img_params(s, r, ro)
+                                r["prim-red-x"] or 0, r["prim-red-y"] or 0,
+                                r["prim-green-x"] or 0, r["prim-green-y"] or 0,
+                                r["prim-blue-x"] or 0, r["prim-blue-y"] or 0,
+                                r["prim-white-x"] or 0, r["prim-white-y"] or 0),
```
All properties are key/value, there is no way to return array. This could be nested property, but it only complicate things.

kasper93 commented on DOCS/man/options.rst:
```diff
@@ -7013,6 +7013,88 @@ them.
+    Default is ``target``. If target display parameters are not available, this
+    will fall back to ``source``. When source metadata is not available, it will
+    fallback to Display P3 primaries and 0.05-1000 cd/m² dynamic range for common
+    consumer display. Note that this is done on individual properties basis,
```
I removed this for now.

kasper93 commented:
> Because mpv currently only gets target_csp metadata at playback initialization, this results in the wrong target colorspace metadata being used in the following scenarios:

That's not true, mpv ask for metadata before each frame render. Though DXGI is caching this and never updating output parameters, which seems like a bug on their side to be honest, because dynamic switching HDR mode is quite natural to support.

I have changed it now to always query current parameters and also added support for getting colorspace with `gpu-api=vulkan`

dyphire commented:
> > Because mpv currently only gets target_csp metadata at playback initialization, this results in the wrong target colorspace metadata being used in the following scenarios:
> 
> That's not true, mpv ask for metadata before each frame render. Though DXGI is caching this and never updating output parameters, which seems like a bug on their side to be honest, because dynamic switching HDR mode is quite natural to support.
> 
> I have changed it now to always query current parameters and also added support for getting colorspace with `gpu-api=vulkan`

This is exactly what I want, after testing it works as expected.

dyphire commented:
I have one last question: what will `target-colorspace-hint=auto` do now on non-Wayland, D3D11 and winvk?

kasper93 commented:
> I have one last question: what will `target-colorspace-hint=auto` do now on non-Wayland, D3D11 and winvk?

Act like `=no`

dyphire commented:
LGTM

Headcrabed commented:
> On some systems this patch for libplacebo may be needed https://code.videolan.org/videolan/libplacebo/-/merge_requests/715 it currently erroneously selects sRGB color space instead of pass through when there is no better match.

What about dmabuf-wayland? As I said in previous PR, wayland part of the code in this PR might won't work at all.... Monitor target csp won't fit with current pl_color_space definition.

To solve this, we might need to use value-based parameter to build up an ICC file and pass that to libplacebo, or make libplacebo able to accept value-based parameter?

Dudemanguy commented:
dmabuf-wayland should just be passthrough like before.

kasper93 commented:
> What about dmabuf-wayland?

dmabuf-wayland doesn't care about any of this. It always passes through the signal and sets the metadata.

> As I said in previous PR, wayland part of the code in this PR might won't work at all.... Monitor target csp won't fit with current pl_color_space definition.
> 
> To solve this, we might need to use value-based parameter to build up an ICC file and pass that to libplacebo, or make libplacebo able to accept value-based parameter?

Both named and parametric primaries are supported. If you mean it's not enough and we need to process an ICC profile to get display characteristics. This is out of the scope of this PR.

Could you tell which composistors currently wouldn't work?

kasper93 commented:
Won't work or will work? I'm confused if they provide value-based primaries then everything is fine.

Dudemanguy commented:
>They don't provide an ICC file or named primaries, and only provide value-based color primaries

Did you say that backwards? Mutter and Kwin should both support named primaries. It's the custom value-based `set_primaries` request that they wouldn't support. We currently don't try that one. It would take some more wayland changes and not sure any compositor uses that yet? Haven't checked lately.

kasper93 commented:
> >They don't provide an ICC file or named primaries, and only provide value-based color primaries
> 
> Did you say that backwards? Mutter and Kwin should both support named primaries. It's the custom value-based `set_primaries` request that they wouldn't support. We currently don't try that one. It would take some more wayland changes and not sure any compositor uses that yet? Haven't checked lately.

If it comes to setting the values. With colorspace-hint it is fully done through Vulkan API. So it's on WSI to set things.

The thing from Wayland that we use is preferred color space. And both value based and named primaries are read and set in our internal structures.

dyphire commented:
Handling and adapting Windows' ACM behavior is a complex operation, which is perhaps beyond the scope of this PR. This is because Windows' ACM has a lot of strange behavior, especially since it is still in an unstable state.

llyyr commented:
> This is not possible. As I said above, monitor EDID spec only defines a single bit to mark whether a monitor is sRGB, if the monitor is not sRGB, then there's only numeric value for monitor's color primaries in EDID data. So compositors can't provide named primaries for monitor's description info.

This doesn't matter, the compositor/WSI can communicate the numerical values for the primaries still through VkHdrMetadataEXT

Headcrabed commented:
> > This is not possible. As I said above, monitor EDID spec only defines a single bit to mark whether a monitor is sRGB, if the monitor is not sRGB, then there's only numeric value for monitor's color primaries in EDID data. So compositors can't provide named primaries for monitor's description info.
> 
> This doesn't matter, the compositor/WSI can communicate the numerical values for the primaries still through VkHdrMetadataEXT

But through this method do we have access to monitors TRC info?

llyyr commented:
> But through this method do we have access to monitor’s TRC info?

That is upto the compositor to parse the edid for it and pass it along through the color management protocol, then mpv can set it for the swapchain

Headcrabed commented:
> > > > But through this method do we have access to monitor’s TRC info?
> > > 
> > > 
> > > That is upto the compositor to parse the edid for it and pass it along through the color management protocol, then mpv can set it for the swapchain
> > 
> > 
> > So in this case, color primaries are set through VkHdrMetadataEXT, but gamma curve is set from color management protocol? Does our current code support this?
> 
> In case of gpu-next, everything is set through Vulkan, both primaries, and transfer function.
> 
> Also, I feel like we are mixing two things in this discussion. One is getting the metadata of the display (or prefered by wayland) the other is setting metadata of the send signal from mpv.

Thanks for the tip. I re-read the code and seems my concern actually doesn't exist, sorry for the noise.

kasper93 commented:
Yes, I know. SDR is not main concern at this moment. For SDR we should use bt.1886 transfer anyway.

kasper93 commented:
Updated for some adjustments in prim selection. Everything is working fine now. Let me know if you see issues.

Headcrabed commented:
Since https://code.videolan.org/videolan/libplacebo/-/merge_requests/716 is refused, what’s the SDR playback behavior for target-colorspace-hint-mode under Wayland now?

mahkoh commented on video/out/wayland_common.c:
```diff
@@ -1651,16 +1655,45 @@ static void info_primaries(void *data, struct wp_image_description_info_v1 *imag
+            closest = i;
+        }
+    }
+    // Allow 0.1 distance
+    if (min_diff > 1000)
+        return;
```
The advantage of using the recommended transfer function is that

1. direct scanout is possible
2. compositing of opaque surfaces can be done with a plain copy

I don't know how much of an advantage choosing an approximate transfer function provides. Maybe it would be better to use the linear transfer function with a floating-point buffer if no exact match exitss.

kasper93 replied:
I can remove that. Frankly if there is no named transfer function provided, we would select bt.1886 for sdr, which is already a good choice.

kasper93 replied:
It's user choice. There is nothing that can say they have properly calibrated monitor to given gamma. bt.1886 is default for sdr.

dyphire replied:
The user can specify it via `target-trc` or icc.

mahkoh replied:
The transfer function chosen by the client has no effect on what is displayed. The choice should be made for efficiency only.

kasper93 replied:
Removed map_tf code.

mahkoh replied:
Isn't scRGB with linear encoding the recommended transfer?

kasper93 replied:
At least on AMD it's not supported, for some reason. On d3d11, on Vulkan it seems to be enumerated. Maybe I will try some day, but it also needs special handling in libplacebo, because it's 80 nits for SDR white (1, 1, 1).

kasper93 commented on video/out/gpu/d3d11_helpers.c:
```diff
@@ -985,23 +987,124 @@ bool mp_d3d11_create_swapchain(ID3D11Device *dev, struct mp_log *log,
+            if (hr == DXGI_ERROR_NOT_FOUND || FAILED(hr))
+                break;
+
+            DXGI_OUTPUT_DESC output_desc;
+            if (SUCCEEDED(IDXGIOutput_GetDesc(pOutput, &output_desc))) {
+                RECT intersection;
+                if (IntersectRect(&intersection, &rect, &output_desc.DesktopCoordinates)) {
+                    LONG area = (intersection.right - intersection.left) *
+                                (intersection.bottom - intersection.top);
+                    if (area > biggest_area) {
+                        biggest_area = area;
+                        SAFE_RELEASE(best_output);
+                        best_output = pOutput;
+                        pOutput = NULL;
+                    }
+                }
+            }
```
> There is no double enumeration. MonitorFromWindow() returns directly the best HMONITOR to use

Do you think `MonitorFromWindow()` does not enumerate monitors to calculate the intersection?

Headcrabed commented:
Sorry to be off-topic, but I’m confused by gpu-next’s SDR TRC curve choice for Vulkan WSI. I saw in libplacebo code that basically all SDR VK_COLOR_SPACE are mapped with BT.1886. What does this mean? Are we using wrong TRC by default?

haasn commented:
I can't test this but the general approach seems reasonable to me.

dyphire commented:
I noticed: when using builds containing this PRs from 2024.07.09 and earlier `--target-colorspace-hint=auto` selects the SDR output for SDR content, whereas from 2024.07.10 onwards `--target-colorspace-hint=auto` transforms the SDR content to PQ output. Is this the expected behavior? 
The reason I ask it is because this PR's current behavior is actually consistent with the current behavior of the master branch, not sure if the behavior from 2024.07.09 and before is normal, although I prefer it.

Headcrabed commented:
> I noticed: when using builds containing this PRs from 2024.07.09 and earlier `--target-colorspace-hint=auto` selects the SDR output for SDR content, whereas from 2024.07.10 onwards `--target-colorspace-hint=auto` transforms the SDR content to PQ output. Is this the expected behavior? The reason I ask it is because this PR's current behavior is actually consistent with the current behavior of the master branch, not sure if the behavior from 2024.07.09 and before is normal, although I prefer it.

Some PRs are just merged on libplacebo side, have you checked them?

kasper93 commented:
> I noticed: when using builds containing this PRs from 2024.07.09 and earlier `--target-colorspace-hint=auto` selects the SDR output for SDR content, whereas from 2024.07.10 onwards `--target-colorspace-hint=auto` transforms the SDR content to PQ output. Is this the expected behavior? The reason I ask it is because this PR's current behavior is actually consistent with the current behavior of the master branch, not sure if the behavior from 2024.07.09 and before is normal, although I prefer it.

You can try latest version, I think I changed it. You can prefer PQ with `--target-trc=pq` for SDR transfers. And I would recommend so, this conversion will be done by you compositor anyway.

dyphire commented:
> Yes this is expected, you are using `=target` mode.

I see, that's what I was wondering, just wanted to make sure it was expected behavior and not a mistake.

Headcrabed commented:
@kasper93 btw I also made a related PR, do you think it’s ok?

https://code.videolan.org/videolan/libplacebo/-/merge_requests/718

kasper93 commented:
> @kasper93 btw I also made a related PR, do you think it’s ok?
> 
> https://code.videolan.org/videolan/libplacebo/-/merge_requests/718

I commented there. Generally I wanted to do the same, but we need to check if it doesn't do weird things with some compositors.
